server/auth/types/openid/routes.ts

```diff
@@ -0,0 +1,213 @@
+          return response.redirected({
+            headers: {
+              location: location,
+            },
+          });
+        }
+
+        // Authentication callback
+        // TODO: figure out a decent way to validate the state from cookie
+        const clientId = this.config.openid?.client_id;
+        const clientSecret = this.config.openid?.client_secret;
+        const query: any = {
+          grant_type: 'authorization_code',
+          code: request.query.code,
+          redirect_uri: `${this.getBaseRedirectUrl()}`,
+          client_id: clientId,
+          client_secret: clientSecret,
+        };
+
+        const requestOptions: any = {
+          payload: stringify(query),
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+          },
+        };
+
+        try {
+          const tokenResponse = await wreck.post(
+            this.openIdAuthConfig.tokenEndpoint as string,
+            requestOptions
+          );
+          if (
+            !tokenResponse.res.statusCode ||
+            tokenResponse.res.statusCode < 200 ||
+            tokenResponse.res.statusCode > 299
+          ) {
+            return response.unauthorized();
+          }
+
+          const tokenPayload: any = this.parseTokenResponse(tokenResponse.payload as Buffer);
+
+          const idToken: string = tokenPayload.id_token;
+          const accessToken: string = tokenPayload.access_token;
+          const refreshToken: string = tokenPayload.refresh_token;
+          const expiresIn: number = tokenPayload.expires_in;
+
+          const user = await this.securityClient.authenticateWithHeader(
+            request,
+            this.openIdAuthConfig.authHeaderName as string,
+            `Bearer ${idToken}`
+          );
+
+          // set to cookie
+          const sessionStorage: SecuritySessionCookie = {
+            username: user.username,
+            credentials: {
+              authHeaderValue: `Bearer ${idToken}`,
+              refresh_token: refreshToken,
+              expires_at: Date.now() + expiresIn * 1000,
+            },
+            authType: 'openid',
+            expiryTime: Date.now() + this.config.cookie.ttl,
+          };
+          this.sessionStorageFactory.asScoped(request).set(sessionStorage);
+          return response.redirected({
+            headers: {
+              location: `${this.core.http.basePath.serverBasePath}/app/kibana`,
+            },
+          });
+        } catch (error) {
+          console.log(error); // TODO: change to logger
+          return response.unauthorized();
+        }
+      }
+    );
+
+    this.router.post(
+      {
+        path: `/auth/logout`,
+        validate: false,
+      },
+      async (context, request, response) => {
+        const cookie = await this.sessionStorageFactory.asScoped(request).get();
+        this.sessionStorageFactory.asScoped(request).clear();
+
+        const token = cookie?.credentials.authHeaderValue.split(' ')[1];
```

Comment:
Minor: probably add a comment about the structure of authHeaderValue so that it is easy to understand the index 1 is corresponding to token.

Reply:
fixed